### PR TITLE
ClassDumper.catchesLinkageError to skip methods without body

### LINKED ISSUE #1352: NPE in ClassDumper
After adding the the linkage-checker-enforcer-rules 1.2.0 to my pom, with:

```
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-enforcer-plugin</artifactId>
                <version>3.0.0-M3</version>
                <dependencies>
                    <dependency>
                        <groupId>com.google.cloud.tools</groupId>
                        <artifactId>linkage-checker-enforcer-rules</artifactId>
                        <version>1.2.0</version>
                    </dependency>
                </dependencies>
                <executions>
                    <execution>
                        <id>enforce-linkage-checker</id>
                        <phase>verify</phase>
                        <goals>
                            <goal>enforce</goal>
                        </goals>
                        <configuration>
                            <rules>
                                <LinkageCheckerRule implementation= "com.google.cloud.tools.dependencies.enforcer.LinkageCheckerRule">
                                    <level>WARN</level>
                                </LinkageCheckerRule>
                            </rules>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
```

I encountered the following exception, immediately after the enforcer plugin gave the following warnings

```
[WARNING] The POM for xerces:xerces-impl:jar:2.6.2 is missing, no dependency information available
[WARNING] The POM for net.sf.ehcache:sizeof-agent:jar:1.0.1 is missing, no dependency information available
[WARNING] The POM for com.ibm.websphere:websphere_uow_api:jar:0.0.1 is missing, no dependency information available
[WARNING] The POM for com.ibm.websphere:uow:jar:6.0.2.17 is missing, no dependency information available
[WARNING] net.sf.ehcache:sizeof-agent:jar:1.0.1 was not resolved. Dependency path: com.xyz:internal-util:jar:1-SNAPSHOT > com.xyz:internal-shaded:jar:1.234 (compile) > net.sf.ehcache:ehcache-core:jar:2.6.11 (compile) > net.sf.ehcache:sizeof-agent:jar:1.0.1 (provided)
```

```
Execution enforce-linkage-checker of goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M3:enforce failed.
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution enforce-linkage-checker of goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M3:enforce failed.
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:148)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.NullPointerException
    at org.apache.bcel.generic.MethodGen.<init> (MethodGen.java:182)
    at com.google.cloud.tools.opensource.classpath.ClassDumper.catchesLinkageError (ClassDumper.java:500)
    at com.google.cloud.tools.opensource.classpath.ClassDumper.catchesLinkageError (ClassDumper.java:518)
    at com.google.cloud.tools.opensource.classpath.LinkageChecker.findSymbolProblem (LinkageChecker.java:482)
    at com.google.cloud.tools.opensource.classpath.LinkageChecker.lambda$findSymbolProblems$1 (LinkageChecker.java:180)
    at com.google.common.collect.ImmutableMultimap.lambda$null$2 (ImmutableMultimap.java:606)
    at java.lang.Iterable.forEach (Iterable.java:75)
    at com.google.common.collect.ImmutableMultimap.lambda$forEach$3 (ImmutableMultimap.java:606)
    at com.google.common.collect.RegularImmutableMap.forEach (RegularImmutableMap.java:185)
    at com.google.common.collect.ImmutableMultimap.forEach (ImmutableMultimap.java:605)
    at com.google.cloud.tools.opensource.classpath.LinkageChecker.findSymbolProblems (LinkageChecker.java:150)
    at com.google.cloud.tools.dependencies.enforcer.LinkageCheckerRule.execute (LinkageCheckerRule.java:203)
    at org.apache.maven.plugins.enforcer.EnforceMojo.execute (EnforceMojo.java:205)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #1352

### CODE REVIEW

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java [hunk 1]
```diff
@@ -480,6 +480,10 @@ boolean catchesLinkageError(String sourceClassName) {
       ClassGen classGen = new ClassGen(sourceJavaClass);
 
       for (Method method : sourceJavaClass.getMethods()) {
+        if (method.getCode() == null) {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java:483]: what sort of methods are these? abstract? native? If so, I'm surprised we haven't hit this before.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java [hunk 2]
```diff
@@ -480,6 +480,10 @@ boolean catchesLinkageError(String sourceClassName) {
...
+          // No need to check the presence of try-catch clause for methods without code
+          continue;
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java:485]: would it be cleaner to simply wrap the loop body in ` if (method.getCode() != null) `?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java (hunk 1 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java:483]: I wrote my investigation in #1352 , which was caused by an invalid class file.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java (hunk 2 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java:485]: If you ask me, I prefer fewer nested code  (less block indentation).

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java (hunk 1 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java:483]: I'd bet this is a BCEL error. The doc for Method.getCode even says "Code attribute of method, if any" so I don't think returning null here is a surprise.

**Pull request merged**